pkg/module/blob/blob.go

```diff
@@ -0,0 +1,315 @@
+const MediaTypeOctetStream = "application/octet-stream"
+
+// Output is the output if read BlobInput.
+type Output struct {
+	Digest string
+	Size   int64
+	Reader io.ReadCloser
+}
+
+type InputType string
+
+const (
+	FileInputType = "file"
+	DirInputType  = "dir"
+)
+
+// Input defines a local resource input that should be added to the component descriptor and
+// to the resource's access.
+type Input struct {
+	// Type defines the input type of the blob to be added.
+	// Note that a input blob of type "dir" is automatically tarred.
+	Type InputType `json:"type"`
+	// MediaType is the mediatype of the defined file that is also added to the oci layer.
+	// Should be a custom media type in the form of "application/vnd.<mydomain>.<my description>"
+	MediaType string `json:"mediaType,omitempty"`
+	// Path is the path that points to the blob to be added.
+	Path string `json:"path"`
+	// CompressWithGzip defines that the blob should be automatically compressed using gzip.
+	CompressWithGzip *bool `json:"compress,omitempty"`
+	// PreserveDir defines that the directory specified in the Path field should be included in the blob.
+	// Only supported for Type dir.
+	PreserveDir bool `json:"preserveDir,omitempty"`
+	// IncludeFiles is a list of shell file name patterns that describe the files that should be included.
+	// If nothing is defined all files are included.
+	// Only relevant for blobinput type "dir".
+	IncludeFiles []string `json:"includeFiles,omitempty"`
+	// ExcludeFiles is a list of shell file name patterns that describe the files that should be excluded from the resulting tar.
+	// Excluded files always overwrite included files.
+	// Only relevant for blobinput type "dir".
+	ExcludeFiles []string `json:"excludeFiles,omitempty"`
+	// FollowSymlinks configures to follow and resolve symlinks when a directory is tarred.
+	// This options will include the content of the symlink directly in the tar.
+	// This option should be used with care.
+	FollowSymlinks bool `json:"followSymlinks,omitempty"`
+}
+
+// Compress returns if the blob should be compressed using gzip.
+func (input Input) Compress() bool {
+	if input.CompressWithGzip == nil {
+		return false
+	}
+	return *input.CompressWithGzip
+}
+
+// SetMediaTypeIfNotDefined sets the media type of the input blob if its not defined
+func (input *Input) SetMediaTypeIfNotDefined(mediaType string) {
+	if len(input.MediaType) != 0 {
+		return
+	}
+	input.MediaType = mediaType
+}
+
+// Read reads the configured blob and returns a reader to the given file.
+func (input *Input) Read(ctx context.Context, fs vfs.FileSystem) (*Output, error) {
+	inputInfo, err := fs.Stat(input.Path)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get info for input blob from %q, %w", input.Path, err)
+	}
+
+	// automatically tar the input artifact if it is a directory
+	if input.Type == DirInputType {
+		if !inputInfo.IsDir() {
+			return nil, fmt.Errorf("resource type is dir but a file was provided")
```

Comment:
```suggestion
			return nil, fmt.Errorf("resource type is dir, but you provided a file")
```
Can we switch to active voice?